core/rawdb/freezer_table.go

```diff
@@ -740,3 +740,28 @@ func (t *freezerTable) dumpIndex(w io.Writer, start, stop int64) {
 	}
 	fmt.Fprintf(w, "|--------------------------|\n")
 }
+
+// forEach iterates every entry in the table serially and in order, calling `fn`
+// with the item as argument. If `fn` returns an error the iteration stops
+// and that error will be returned.
+func (t *freezerTable) forEach(fn func(uint64, []byte) error) error {
+	items := atomic.LoadUint64(&t.items)
```

Comment:
Would be neat to somehow prevent write operations while `forEach` is executing. `t.RetrieveItems` is safe, in itself, but since we're iterating over a wider range, it would be even better to have a mutex guarantee across the entire range, not just per-batch. 
Not sure how to solve that. 

Incidentally: this `forEach` function doesn't have to be part of `freezerTable`, right? It could be implemented on the caller-side too, couldn't it? Then it would be protected by the freezer writelock, and not exposed to any other callers. 

Just speculatng here...

Reply:
Yes a mutex guarantee is kinda tricky here since it possibly takes a long time. I made it a method of `freezerTable` thinking it might be useful for other purposes where there isn't concurrent access to the table (e.g. `InitDatabaseFromFreezer`). For now I changed it into an anonymous function defined in `MigrateTable`. There its clear that no other goroutine is accessing the table and the `writelock` is also taken.

In the meanwhile I'll think over how we can handle locking better

Reply:
I think it's fine as you've done now, having it anonymous and protected by `f.writeLock`.